File: Palindrome/palindrome.go
package palindrome

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	middle, fast := head, head

	for fast != nil {
		middle = middle.Next
		fast = fast.Next

		if fast != nil {
			fast = fast.Next
		}
	}

	var tail *ListNode
	for middle != nil {
		next := middle.Next
		middle.Next = tail
		tail = middle
		middle = next
	}

	for tail != nil {
		if head.Val != tail.Val {
			return false
		}
		head = head.Next
		tail = tail.Next
	}

	return true
}

/*
func nodetoSlice(head *ListNode) []int {
	result := []int{}
	for head != nil {
		result = append(result, head.Val)
		head = head.Next
	}

	return result
}
*/

File: Palindrome/palindrome.cs
public class ListNode
{
    public int val;
    public ListNode next;

    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}

public class Solution
{
    public bool IsPalindrome(ListNode head)
    {
        var slow = head;
        var fast = head;
        ListNode prev = null;

        while (fast != null && fast.next != null)
        {
            fast = fast.next.next;
            ListNode next = slow.next;
            slow.next = prev;
            prev = slow;
            slow = next;
        }

        if (fast != null)
        {
            slow = slow.next;
        }

        while (slow != null)
        {
            if (slow.val != prev.val)
            {
                return fals;
            }

            slow = slow.next;
            prev = prev.next;
        }

        return true;
    }
}
